entoas/extension.go

```diff
@@ -0,0 +1,183 @@
+// Annotations of the extensions.
+func (ex *Extension) Annotations() []entc.Annotation {
+	return []entc.Annotation{ex.config}
+}
+
+// DefaultPolicy sets the default ExclusionPolicy to use of none is given on a (sub-)schema.
+func DefaultPolicy(p policy.Policy) ExtensionOption {
+	return func(ex *Extension) error {
+		ex.config.DefaultPolicy = p
+		return nil
+	}
+}
+
+// Mutations adds the given mutations to the spec generator.
+func Mutations(ms ...MutateFunc) ExtensionOption {
+	return func(ex *Extension) error {
+		ex.mutations = append(ex.mutations, ms...)
+		return nil
+	}
+}
+
+// SimpleModels enables the simple model generation feature.
+func SimpleModels() ExtensionOption {
+	return func(ex *Extension) error {
+		ex.config.SimpleModels = true
+		return nil
+	}
+}
+
+// SpecTitle sets the title of the Info block.
+func SpecTitle(v string) ExtensionOption {
+	return Mutations(func(_ *gen.Graph, spec *spec.Spec) error {
+		spec.Info.Title = v
+		return nil
+	})
+}
+
+// SpecDescription sets the title of the Info block.
+func SpecDescription(v string) ExtensionOption {
+	return Mutations(func(_ *gen.Graph, spec *spec.Spec) error {
+		spec.Info.Description = v
+		return nil
+	})
+}
+
+// SpecVersion sets the version of the Info block.
+func SpecVersion(v string) ExtensionOption {
+	return Mutations(func(_ *gen.Graph, spec *spec.Spec) error {
+		spec.Info.Version = v
+		return nil
+	})
+}
+
+// DumpTo dumps the current specs content to the given io.Writer.
+func DumpTo(out io.Writer) ExtensionOption {
+	return func(ex *Extension) error {
+		ex.out = out
+		return nil
+	}
+}
+
+// generator returns a gen.Hook that creates a Spec for the given gen.Graph.
+func (ex *Extension) generate(next gen.Generator) gen.Generator {
+	return gen.GenerateFunc(func(g *gen.Graph) error {
+		// Let ent create all the files.
+		if err := next.Generate(g); err != nil {
+			return err
+		}
+		// Spec stub to fill.
+		s := &spec.Spec{
+			Info: &spec.Info{
+				Title:       "Ent Schema API",
+				Description: "This is an auto generated API description made out of an Ent schema definition",
+				Version:     "0.0.0",
+			},
+			Components: &spec.Components{
+				Schemas:    make(map[string]*spec.Schema),
+				Responses:  make(map[string]*spec.Response),
+				Parameters: make(map[string]*spec.Parameter),
+			},
+		}
+		// Run the generator.
+		if err := generate(g, s); err != nil {
+			return err
+		}
+		// Run the user provided mutations.
+		for _, m := range ex.mutations {
+			if err := m(g, s); err != nil {
+				return err
+			}
+		}
+		// Dump the spec.
+		b, err := json.MarshalIndent(s, "", "  ")
+		if err != nil {
+			return err
+		}
+		if ex.out != nil {
+			_, err = ex.out.Write(b)
+			return err
+		}
+		return os.WriteFile(filepath.Join(g.Target, "openapi.json"), b, 0644)
+	})
+}
+
+// Name implements entc.Annotation interface.
+func (c Config) Name() string {
+	return "entopaConfig"
+}
+
+// Decode from ent.
+func (c *Config) Decode(o interface{}) error {
```

Comment:
Does this need to be public ?

Reply:
Not sure. TBH I copied this from Ent. @a8m is the original author of the `Decode` method. But I guess you are right, this one can be unexported.